### ISSUE TITLE
Enables me to use software developed with the PySimpleGUI package

### ISSUE THREAD

**Issue by c469591** [nvaccess/nvda#15016]
### Is your feature request related to a problem? Please describe.

no

### Describe the solution you'd like

Allow me to smoothly operate software developed using the PySimpleGUI library.

### Describe alternatives you've considered

no

### Additional context

I hope to be able to smoothly use software developed using PySimpleGUI for GUI, 
for example UVR →  https://github.com/Anjok07/ultimatevocalremovergui.
Nowadays, more and more software is being developed using PySimpleGUI instead of WxPython, 
which means that I am completely unable to operate an increasing number of software applications using NVDA.
Thank you!

**Comment by XLTechie** [comment]
I believe it's QT based. QT has been very slow to implement accessibility, and I 
don't know how well PySimpleGUI implements even the accessible parts of QT.

Complain to the PySimpleGUI developers about this.

**Comment by mzanm** [comment]
PySimpleGUI uses Tkinter which is way worse than Qt. It's completely, fully inaccessible. NVDA reads nothing.

**Comment by c469591** [comment]
As far as I know, ZDSR, a screen reader software from China, has already supported the operation of this GUI. Through ZDSR, it is possible to operate UVR and other similar software. Therefore, I believe that NVDA should also find a way to achieve the same functionality.

**Comment by c469591** [comment]
In addition, this new feature is currently only implemented in the beta version of ZDSR, the effect is very good

**Comment by josephsl** [comment]
Hi,

This is an interesting data point - perhaps ZDSR might be using a recognition engine or something to recognize and work with elements drawn to the screen for which accessibility API information is not defined (or if defined, done incorrectly). The ideal fix should come from ap developers, more specifically developers of GUI toolkits as exposing accessibility information can benefit more than one screen reader (a good example is QT where toolkit developers realized benefits of improving accessibility at the toolkit level). As others noted (directly or not), I advise getting app developers involved and explain the benefits of using a more accessible GUI toolkit and/or strategies to persuade Tkinter folks about accessibility.

Thanks.

**Comment by josephsl** [comment]
Hi, let us not forget Narrator, as Narrator has become a “gold standard” (with qualifiers) on Windows 10 and later. Thanks.

**Comment by c469591** [comment]
I took UVR to test, the windows narrator can not read aloud any controls on UVR

**Comment by seanbudd** [comment]
We are unable to process this issue as it stands.

This is a problem with the implementation of the framework, as a starting point please submit feedback to the authors to let them know about the problem.

Kind Regards,
NV Access Software Developers

**Comment by Emil-18** [comment]
I created a dll that hooks ExtTextOutW, and prints the text given to it to a file. I injected the dll into idle, an application written in tkinter, and got this result. Of note is that NVDA's display model only finds the text of the menu bar.
If NVDA's display model gets improved in such a way that it can find the text in tkinter applications, it is at least possible to use them.
[idleText.txt](https://github.com/nvaccess/nvda/files/15000139/idleText.txt)

**Comment by Adriani90** [comment]
The finding by @Emil-18 in the last comment is quite interesting. Maybe @LeonarddeR has some suggestions here as well.

**Comment by LeonarddeR** [comment]
The display model code of NVDA is pretty legacy and pretty complex. When you fix or improve one thing, you are pretty likely to break another.

**Comment by Emil-18** [comment]
@Adriani90, @LeonarddeR I did actualy find out why NVDA can't read the text in tkinter applications with its display model. Tkinter, for what ever reason, draws its text to a memory DC, then detach that dc from the bitmap, then attach a new memory dc to the same bitmap, and finaly bitblits the text to the screen using BitBlt. NVDA doesn't track the text when a bitmap swaps dcs like that, it just deletes it. Allthough I managed to make it work somewhat, i still have a problem where random bits of text in tkinter applications are not rendered by the display model. This happens to random text at random times, for example, when testing it with idle, i can find the >>> text, but if I redraw the screen, the >>> text may be gone. If I redraw again, it may be gone or visible

**Comment by LeonarddeR** [comment]
@Emil-18 Interesting to see that you investigated this so deeply. I've seen several related results in other applications as well. Minimizing and maximizing a window can make things even worse.

**Comment by ABuffEr** [comment]
> somewhat, i still have a problem where random bits of text in tkinter applications are not rendered by the display model.

Have you way to use SWP_NOCOPYBITS? As I said on nvda-addons, I saw exactly what you describe using redraw, when I wrote refreshTextContent function [here.](https://raw.githubusercontent.com/ABuffEr/labelAutofinderCore/refs/heads/main/utils.py)

**Comment by Emil-18** [comment]
@ABuffEr The problem isn't that you have to redraw the text to make the display model render it. It is that when you redraw once, some text are visible and some aren't. When you redraw again, parts of the text that was visible are now hidden and vice versa, And so on

**Comment by ABuffEr** [comment]
> It is that when you redraw once, some text are visible and some aren't. When you redraw again, parts of the text that was visible are now hidden and vice versa, And so on

Yes, but it seems very similar to what I observed if I use SetWindowPos without SWP_NOCOPYBITS flag. So I suggested it.

**Comment by ABuffEr** [comment]
@Emil-18, surely in [RadioBOSS](https://www.djsoft.net), inside Music Library (ctrl+b) or Track Tool if I remember well, and maybe in [WinSCP](https://winscp.net/eng/index.php) too.

**Comment by Emil-18** [comment]
@Adriani90, @ABuffEr, @LeonarddeR I fixed THE ISSUE WITH THE TEXT NOT SHOWING AT RANDOM TIMES. I modified NVDA so it stores memory display models via bitmap pointers, instead of memory DCs, as all text written with memory DCs, is written to  a bitmap, instead of the memory dc. I tested it with idle, and have attached text files where I recorded NVDA's speech both for the normal version, and the modified version.

[normalNVDAIdleText.txt](https://github.com/user-attachments/files/21606266/normalNVDAIdleText.txt)

[Modified NVDA idle text.txt](https://github.com/user-attachments/files/21606263/Modified.NVDA.idle.text.txt)

This technique also makes the display model find more text in Eclipse, and possibly more applications as well, however, it causes text in other areas, such as in the run dialog, and in NVDA's own interface, to be scrambled. To fix this, I clear the rectangle copied from the source model in StretchBlt_helper, if it is a memory model and if it isn't equal to the dest model. This causes text in yet other areas to disappear, such as the text "MySQL Connections" in the main window of MySql workbench. Curiously enough, JAWS is missing that exact text as well.

Note that there are still text that JAWS and ZDSR finds that NVDA doesn't find.

I think NVDA should do the following.

* Let NVDA's display model use its own dll, rather than combine it with the dll used to implement virtual buffers, so new display models can be added, without worrying about having to keep both virtual buffers up to date.
* Add another display model to NVDA, in yet another dll, that implements this, or something similar/better. Let the original display model stay exactly as it is.
* Add support for graphics and some stuff drawn with the "DrawThemeBackground" function, such as check marks, in the new display model. Do not show this in obj.displayText. Allow the user to configure if this stuff should be shown when navigating the DisplayModelTextInfo. JAWS has support for this.
* Let the user choose between the display models. Both JAWS and ZDSR has two display models that the user can choose between. JAWS has its graphicks driver, and GDI hooking. Both of ZDSR's display models are using GDI hooking I think.
* Allow add-ons to easily add their own display models.

modified display model code:

[gdiHooks.zip](https://github.com/user-attachments/files/21606844/gdiHooks.zip)

**Comment by seanbudd** [comment]
Closing as won't fix

The project is abandoned as of Jan 2026 https://github.com/PySimpleGUI/PySimpleGUI

Small, specific issues which could be fixed in NVDA can be reopened for anything remaining we can do
